### PR TITLE
MDEV-28127 SIGSEGV in row_mysql_store_col_in_innobase_format, and Ass…

### PR DESCRIPTION
**Pull request by TakeKiyo**
- [x] *The Jira issue number for this PR is: MDEV-28127*

## Description
MDEV-28127 SIGSEGV in row_mysql_store_col_in_innobase_format, and Assertion `pos < table->n_v_def` failed in dict_table_get_nth_v_col, and Assertion `table->s->db_options_in_use == part_table->s->db_options_in_use` failed in compare_table_with_partition

With InnoDB, executing an exchange partition that has a virtual column and a table that doesn't have a virtual column was allowed. Then when trying to insert a value to the table,
Assertion `pos < table->n_v_def'` failed.

With MyISAM, when trying to execute an exchange partition that has a virtual column and a table that doesn't have a virtual column,
Assertion `table->s->db_options_in_use == part_table->s->db_options_in_use'` failed.

However, these executions should not be allowed since a normal column and a virtual column (a constant one at that) are not really interchangeable.

To fix the problem, I added the code that confirms the number of virtual columns is equal when executing exchange partition. 

## How can this PR be tested?
The test cases are included in pull request.
I added code that test InnoDB, MyISAM, Aria.

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [x] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

### CODE REVIEW

**Comment by TakeKiyo** [comment]
Hi @FooBarrior,
I created this pull request 10 days ago. 

I’m sorry to bother you, but I would be grateful if you could check this pull request.

**Comment by FooBarrior** [comment]
@TakeKiyo Hi! Yes, I've been hurrying with an overdue on another task. Sorry that I couldn't take a look earlier. I'll start revierwing tomorrow, but yours one is the second one in the queue ;-)

**Review by FooBarrior — CHANGES_REQUESTED** [review]
@TakeKiyo Hello Kiyoshi!
> To fix the problem, I added the code that confirms the number of virtual columns is equal when executing exchange partition.

Why would we care about a number of virtual columns? It seems, that virtual columns is the thing that doesn't impact on prtitions binary compatibility (unless they are indexed...).

Can we just check that a number of *stored* columns equals? Note that virtual columns can be stored, too. I reckon there should be a corresponding field in TABLE_SHARE.

Also, as we remembered virtual column indexes here, can you check, what happens on them?
E.g. let's say one column is just virtual, and another one is indexed virtual column.

File: mysql-test/main/partition_exchange.test [hunk 1]
```diff
@@ -554,3 +554,31 @@ ALTER TABLE t1 EXCHANGE PARTITION p0 WITH TABLE t2;
 
 # Cleanup
 DROP TABLE t1, t2;
+
+--echo #
+--echo # MDEV-28127 SIGSEGV in row_mysql_store_col_in_innobase_format, and Assertion `pos < table->n_v_def' failed in dict_table_get_nth_v_col, and Assertion `table->s->db_options_in_use == part_table->s->db_options_in_use' failed in compare_table_with_partition
```
[the remark this answers is not in the document]
Comment (TakeKiyo) [reply mysql-test/main/partition_exchange.test:559]: OK, thank you for your pointing it out.

**Comment by TakeKiyo** [comment]
@FooBarrior Hi, Nikita.
Thank you for checking this pull request.

> Why would we care about a number of virtual columns? 

I added the code to check the number of virtual columns since in the [Jira](https://jira.mariadb.org/browse/MDEV-28127?focusedCommentId=217376&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-217376), Marko commented 

`The operation should be rejected, because a normal column and a virtual column (a constant one at that) are not really interchangeable.`

> I reckon there should be a corresponding field in TABLE_SHARE.

I found the field `stored_field` in TABLE_SHARE.
I’ll revise the code to check the number of *stored* columns equals based on your comment.

> E.g. let's say one column is just virtual, and another one is indexed virtual column.

I don't think I fully understand your comment.
Should I check what will happen if I create the following two tables?

```
CREATE TABLE t1 (a INT, b INT GENERATED ALWAYS AS (1) VIRTUAL, KEY(a)) ENGINE=InnoDB;
CREATE TABLE t2 (a INT, b INT GENERATED ALWAYS AS (1) VIRTUAL, KEY(b)) ENGINE=InnoDB;
```

**Comment by TakeKiyo** [comment]
@FooBarrior Hi, Nikita.

I appreciate it if you could check my comment.

**Comment by FooBarrior** [comment]
Hi Takeda!
Sorry, I saw you made a commit after your comment, so I thought you're still having a work in progress.

`stored_field` looks coorrect to use, yes.

Now about indexes. We have an ability to make indexes on non-stored virtual columns. This means that the column will not be stored in a clustered index (or a myisam record), but the index will be created. The index itself is stored:)

So i think such partitions are also not interchangeable. I wonder if it is forbidden now. Can you check?

>Should I check what will happen if I create the following two tables?
> `CREATE TABLE t1 (a INT, b INT GENERATED ALWAYS AS (1) VIRTUAL, KEY(a)) ENGINE=InnoDB;`
> `CREATE TABLE t2 (a INT, b INT GENERATED ALWAYS AS (1) VIRTUAL, KEY(b)) ENGINE=InnoDB;`

Yes, i think this example will be a good one, also I'd check two tables both with indexed vcols, but of incompatible types (like INT and BYTE)

**Comment by TakeKiyo** [comment]
@FooBarrior Hi,Nikita.
Thank you for checking my comment and I am sorry for my misleading behavior.

> stored_field looks coorrect to use, yes.

Thank you for confirming.

I created 4 test cases based on your comment:

1. With InnoDB, one column is just virtual, and another one is indexed virtual column.
```
CREATE TABLE t1 (a INT, b INT GENERATED ALWAYS AS (1) VIRTUAL, KEY(a)) ENGINE=InnoDB PARTITION BY RANGE (a) (PARTITION p VALUES LESS THAN (1));
CREATE TABLE t2 (a INT, b INT GENERATED ALWAYS AS (1) VIRTUAL, KEY(b)) ENGINE=InnoDB;
```
2. With MyISAM, one column is just virtual, and another one is indexed virtual column.
```
CREATE TABLE t1 (a INT, b INT GENERATED ALWAYS AS (1) VIRTUAL, KEY(a)) ENGINE=MyISAM PARTITION BY RANGE (a) (PARTITION p VALUES LESS THAN (1));
CREATE TABLE t2 (a INT, b INT GENERATED ALWAYS AS (1) VIRTUAL, KEY(b)) ENGINE=MyISAM;
```
3.  With InnoDB, two tables both with indexed vcols, but of incompatible types (INT and BINARY).
```
CREATE TABLE t1 (a INT, b INT GENERATED ALWAYS AS (1) VIRTUAL, KEY(b)) ENGINE=InnoDB PARTITION BY RANGE (a) (PARTITION p VALUES LESS THAN (1));
CREATE TABLE t2 (a INT, b BINARY GENERATED ALWAYS AS ('1') VIRTUAL, KEY(b)) ENGINE=InnoDB;
```
4. With MyISAM, two tables both with indexed vcols, but of incompatible types (INT and BINARY).
```
CREATE TABLE t1 (a INT, b INT GENERATED ALWAYS AS (1) VIRTUAL, KEY(b)) ENGINE=MyISAM PARTITION BY RANGE (a) (PARTITION p VALUES LESS THAN (1));
CREATE TABLE t2 (a INT, b BINARY GENERATED ALWAYS AS ('1') VIRTUAL, KEY(b)) ENGINE=MyISAM;
```

I executed the 
`ALTER TABLE t1 EXCHANGE PARTITION p WITH TABLE t2;`
 for all patterns and it resulted in the same error 
`ERROR HY000: Tables have different definitions.`.

So, I suppose the partitions you pointed out are already not interchangeable.

**Comment by TakeKiyo** [comment]
@FooBarrior Hi,Nikita.
I am grateful if you check my comment.
Thank you in advance.

**Comment by FooBarrior** [comment]
Hi @TakeKiyo! Thanks for notifying. This time i really missed your message before, sorry.

I went through the test you've made.

> CREATE TABLE t1 (a INT, b INT GENERATED ALWAYS AS (1) VIRTUAL, KEY(a)) ENGINE=InnoDB PARTITION BY RANGE (a) (PARTITION p VALUES LESS THAN (1));
> CREATE TABLE t2 (a INT, b INT GENERATED ALWAYS AS (1) VIRTUAL, KEY(b)) ENGINE=InnoDB;
Ok, so we see, that a number of keys matching is not enough, they (likely) should stay on the same positions. Good for beginning.
In this example we learn that keys should be on same columns. But i was wondering, why EXCHANGE PARTITIONS from the failing test was passing. Somehow, `mysql_compare_tables` considered those correct. And I had a feeling that just comparing stored_fields is not enough.
So I made some digging of my own.

```sql
create table t1 (b int, a int as(1), key(a), c smallint as (1)) partition by range (a) (partition p values less than (1));
create table t2 (b int, a int as (1),key(a));
ALTER TABLE t1 EXCHANGE PARTITION p WITH TABLE t2;
```
doesn't work. Ok we need virtual columns to match, too.
```sql
create table t1 (b int, a int as(1), key(a), c smallint as (1)) partition by range (a) (partition p values less than (1));
create table t2 (b int, a int as (1),key(a), c int as (1));
ALTER TABLE t1 EXCHANGE PARTITION p WITH TABLE t2;
```
Same. Their types should match, too.

```sql
create table t1 (b int, a int, key(a), c int as (1)) partition by range (a) (partition p values less than (1));
create table t2 (b int, a int as (1),key(a), c int);
ALTER TABLE t1 EXCHANGE PARTITION p WITH TABLE t2;
```
Worked! Ok let's see if inserts will fail
```sql
insert into t1(b, a) values (0, -1);
insert into t2(b, c) values (2, 4);
```
nope, works just fine.
```sql
select * from t2 where c > 3;
```
also. It seems that `key(a)` from t1 now corresponds to `c` in t2, which is, however, a third column.

Then, what if we'll change a column `c` to the one not matching the key width?
```sql
create table t1 (b int, a int, key(a), c smallint as (1)) partition by range (a) (partition p values less than (1));
create table t2 (b int, a int as (1),key(a), c smallint);
ALTER TABLE t1 EXCHANGE PARTITION p WITH TABLE t2;
insert into t1(b, a) values (0, -1);
insert into t2(b, c) values (2, 4);
```
boom! It worked. So it seems that `mysql_compare_tables` just doesn't check the columns' virtualness when it compares metadata.

Basically, that's the investigation i was waiting from you. To make a correct fix, you should first find out, why it _really_ fails.

Now remove, the columns number comparison, get into `mysql_compare_tables` and correct the columns comparison.
Good luck!

File: mysql-test/main/partition_exchange.test [hunk 2]
```diff
@@ -554,3 +554,34 @@ ALTER TABLE t1 EXCHANGE PARTITION p0 WITH TABLE t2;
...
+
+--echo #
+--echo # MDEV-28127 SIGSEGV in row_mysql_store_col_in_innobase_format, and Assertion  
+--echo # `pos < table->n_v_def' failed in dict_table_get_nth_v_col, and Assertion
+--echo # `table->s->db_options_in_use == part_table->s->db_options_in_use' failed 
+--echo # in compare_table_with_partition
+--echo #
+SET sql_mode='';
+CREATE TABLE t1 (a INT,KEY(a)) ENGINE=InnoDB PARTITION BY RANGE (a) (PARTITION p VALUES LESS THAN (1));
+CREATE TABLE t2 (a INT GENERATED ALWAYS AS (1) VIRTUAL,KEY(a)) ENGINE=InnoDB;
```
Comment (dr-m) [inline mysql-test/main/partition_exchange.test:566]: I think that it would be good to cover `GENERATED…STORED` in addition to `GENERATED…VIRTUAL`.

**Comment by TakeKiyo** [comment]
Thank you for your investigation and advice.
I will deep dive to `mysql_compare_tables` and try to correct the columns comparison.

File: mysql-test/main/partition_exchange.test (hunk 2 above)
Comment (TakeKiyo) [reply mysql-test/main/partition_exchange.test:566]: Thank you for pointing that out.
I will add code to cover `GENERATED…STORED`.

**Comment by TakeKiyo** [comment]
@FooBarrior Hi, Nikita.

I want to make sure my understanding of your comment is correct.

You use MariaDB client to run the test, which means the storage engine is InnoDB?

I ran the following test code in my environment using MTR, and it didn't work. 

```
--source include/have_partition.inc
--source include/have_innodb.inc

create table t1 (b int, a int, key(a), c smallint as (1)) Engine=InnoDB partition by range (a) (partition p values less than (1));
create table t2 (b int, a int as (1),key(a), c smallint)  Engine=InnoDB;
ALTER TABLE t1 EXCHANGE PARTITION p WITH TABLE t2;
insert into t1(b, a) values (0, -1);
insert into t2(b, c) values (2, 4);
```

result:
```
parts.example 'innodb'     [ fail ]
        Test ended at 2022-08-01 06:46:01

CURRENT_TEST: parts.example
mysqltest: At line 7: query 'insert into t1(b, a) values (0, -1)' failed: 2013: Lost connection to MySQL server during query
```


I wonder why the test code worked in your environment and it did not in my environment.

**Comment by FooBarrior** [comment]
No further investigation was made, so I'm closing this PR. If a submitter is free to continue here, once they'll have new progress.

**Comment by mariadb-RoelVandePaar** [comment]
@TakeKiyo What was the reason for the 2013 lost connection? A new crash?
